api/rosetta/data_transaction_integration_test.go

```diff
@@ -0,0 +1,554 @@
+// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+// +build integration
+
+package rosetta_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/labstack/echo/v4"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/api/rosetta"
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/rosetta/configuration"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/meta"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+func TestGetTransaction(t *testing.T) {
+
+	db := setupDB(t)
+	api := setupAPI(t, db)
+
+	var (
+		firstHeader      = knownHeaders(44)
+		multipleTxHeader = knownHeaders(165)
+		lastHeader       = knownHeaders(181)
+	)
+
+	const (
+		firstTx = "d5c18baf6c8d11f0693e71dbb951c4856d4f25a456f4d5285a75fd73af39161c"
+
+		// two transactions in a single block
+		firstOfTwoTx  = "23c486cfd54bca7138b519203322327bf46e43a780a237d1c5bb0a82f0a06c1d"
+		secondOfTwoTx = "3d6922d6c6fd161a76cec23b11067f22cac6409a49b28b905989db64f5cb05a5"
+
+		lastTx = "780bafaf4721ca4270986ea51e659951a8912c2eb99fb1bfedeb753b023cd4d9"
+	)
+
+	tests := []struct {
+		name string
+
+		request              rosetta.TransactionRequest
+		validateTransactions transactionValidationFn
+	}{
+		{
+			name:                 "some cherry picked transaction",
+			request:              requestTransaction(firstHeader, firstTx),
+			validateTransactions: validateTransfer(t, firstTx, "754aed9de6197641", "631e88ae7f1d7c20", 1),
+		},
+		{
+			name:                 "first in a block with multiple",
+			request:              requestTransaction(multipleTxHeader, firstOfTwoTx),
+			validateTransactions: validateTransfer(t, firstOfTwoTx, "8c5303eaa26202d6", "72157877737ce077", 100_00000000),
+		},
+		{
+			name:                 "second in a block with multiple",
+			request:              requestTransaction(multipleTxHeader, secondOfTwoTx),
+			validateTransactions: validateTransfer(t, secondOfTwoTx, "89c61aa64423504c", "82ec283f88a62e65", 1),
+		},
+		{
+			name:                 "last transaction recorded",
+			request:              requestTransaction(lastHeader, lastTx),
+			validateTransactions: validateTransfer(t, lastTx, "668b91e2995c2eba", "89c61aa64423504c", 1),
+		},
```

Comment:
I guess that here we probably want to add a test case for getting the first transaction within a block, and one to get the last within a block with multiple transactions

Reply:
That's what the `first in a block with multiple` and `second in a block with multiple` does.

Unfortunately I had no blocks with more than two transactions, so `first` and `second` are actually `first` and `last`. Or the best I could do. I could probably add a comment to clarify that

Reply:
Ah okay yeah I didn't get that. From IIRC it's supposed to be start middle and end, so maybe we need at least 3-4 elements for a proper test? Would it be too inconvenient/verbose?

Reply:
It wouldn't be too verbose, but I'd need to recreate the flow protocol state database and generate more events using `flow-sim`, and even then I'm not sure when I'd get the stars to align and have >=3 transactions to end up in the same block. Then I'd need to create a new index snapshot and update it.

It all depends how important it is to have that case covered. It does feel like it would be nice.

@awfm9 WDYT?

Reply:
I think it's not necessary; the Flow Sim will eventually create large amounts of transactions and we will have dozens per block at times, but it will also drastically increase the size of the test snapshot.